docs/UnivBot.man
.\" Man page for UnivBot
.TH UnivBot 1 "2013-15-01" "Linux" "UnivBot IRC Bot"

.SH NAME
UnivBot - A highly customizable Internet Relay Chat bot

.SH SYNOPSIS
.\" Options to actually run / quit
.B UnivBot
.RB [ \-q | \--quit ]
.RB [ \-k | \--kill ]
.RB [ \-d | \--fork ]
.RB [ \--quiet ]
.RB [ -c
.IR filename
.RB |  \--config=
.IR filename ]
.\" Options to configure or get help
.br
.B UnivBot
.RB [ \-h | \--help ]
.br
.B UnivBot
.RB [ \-m | \--migrate ]
.br
.B UnivBot
.RB [ \-w | \--configure-all ]
.RB [ \--configure-modules ]
.RB [ \--configure-database ]

.SH DESCRIPTION
.B UnivBot
runs a bot that provides some usefull (and some useless) commands in one or more
Internet Relay Chat channels. It reads configuration options, such as the
network to connect to, channels to join, and more from a config file specified
by the 
.B -c
option. If no config file is specified,
.B ~/.UnivBot/default.cfg
is used. If the configuration file being used does not exist, a brief
configuration wizard will be run to set up the basic options, and optionally set
up a database for user data and any modules that may need configuration.
.P
In most cases, simply running 
.B UnivBot
will be sufficient; the bot will start after initial configuration, and will run
until it is told to quit through IRC.

.SH OPTIONS
.TP 5
\-h or --help
Show a brief summary of these options.
.TP
\-c
.IR filename
or --config=
.IR filename
.br
Run UnivBot with the configuration options from the specified file. An absolute
pathname can be given, or the name of a file within the configuration directory
may also be given.
.B ~/.UnivBot
is the preferred location, but this and 
.B -c
will also look in 
.B ~/.jenni
and 
.B ~/.phenny
for legacy reasons. When the suffix of the file is
.B .cfg
, it may be omitted.
.TP
\-d or --fork
Run UnivBot in the background. The bot will safely run in the background, and can
later be killed with 
.B -q
or 
.B -d
\.
.TP
\-q or --quit
Gracefully quit a daemonized instance of UnivBot. This option determines the 
instance to quit by the name of the config file given by the 
.B -c
option, if used, so the same argument must be given with 
.B -c
when using 
.B -q.
.TP
\-k or --kill
Terminates a daemonized instance of UnivBot. This option determines the instance
to kill by the name of the config file given by the 
.B -c
option, if used, so the same argument must be given with 
.B -c
when using 
.B -k.
.TP
\-l or --list
List all config files found in the usual configuration file directories:
.B ~/.UnivBot ~/.jenni
and 
.B ~/.phenny
.TP
\-m or --migrate
Attempt to migrate a configuration file for an older version of the bot to the 
new format. Versions prior to 3.1 used a configuration file written in Python,
which can be converted with this tool.
.TP
\--quiet
Suppress all output from the bot to the terminal.
.TP
\-w or --configure-all
Run the initial configuration wizard. This can be run with
.B -c
to configure a specified file, or without it to work on the default file.
.TP
\--configure-modules
Check for modules with configuration options, and allow the user to modify those
options. Changes will be added to the default file, or the file given to
.B -c
if used.
.TP
\--configure-database
Run a wizard for configuring the user information database options. Changes
will be added to the default file, or the file given to
.B -c
if used.


UnivBot.py
#!/usr/bin/env python2.7
# coding=utf8

from __future__ import unicode_literals
from __future__ import print_function

import sys
from UnivBot.tools import stderr

if sys.version_info < (2, 7):
    stderr('Error: Requires Python 2.7 or later. Try python2.7 UnivBot')
    sys.exit(1)
if sys.version_info.major == 3 and sys.version_info.minor < 3:
    stderr('Error: When running on Python 3, Python 3.3 is required.')
    sys.exit(1)

import os
import argparse
import signal

from UnivBot.__init__ import run
from UnivBot.config import Config, create_config, ConfigurationError, wizard
import UnivBot.tools as tools
import UnivBot.web

homedir = os.path.join(os.path.expanduser('~'), '.UnivBot')

def enumerate_configs(extension='.cfg'):
    configfiles = []
    if os.path.isdir(homedir):
        UnivBot_dotdirfiles = os.listdir(homedir)  # Preferred
        for item in UnivBot_dotdirfiles:
            if item.endswith(extension):
                configfiles.append(item)

    return configfiles


def find_config(name, extension='.cfg'):
    if os.path.isfile(name):
        return name
    configs = enumerate_configs(extension)
    if name in configs or name + extension in configs:
        if name + extension in configs:
            name = name + extension

    return os.path.join(homedir, name)


def main(argv=None):
    global homedir
    # Step One: Parse The Command Line
    try:
        parser = argparse.ArgumentParser(description='DreamBot, a IRC Bot',
                                         usage='%(prog)s [options]')
        parser.add_argument('-c', '--config', metavar='filename',
                            help='use a specific configuration file')
        parser.add_argument("-d", '--fork', action="store_true",
                          dest="deamonize", help="Deamonize UnivBot")
        parser.add_argument("-q", '--quit', action="store_true", dest="quit",
                          help="Gracefully quit UnivBot")
        parser.add_argument("-k", '--kill', action="store_true", dest="kill",
                          help="Kill UnivBot")
        parser.add_argument('--exit-on-error', action="store_true",
                          dest="exit_on_error", help=(
                              "Exit immediately on every error instead of "
                              "trying to recover"))
        parser.add_argument("-l", '--list', action="store_true",
                          dest="list_configs",
                          help="List all config files found")
        parser.add_argument("-m", '--migrate', action="store_true",
                          dest="migrate_configs",
                          help="Migrate config files to the new format")
        parser.add_argument('--quiet', action="store_true", dest="quiet",
                          help="Supress all output")
        parser.add_argument('-w', '--configure-all', action='store_true',
                          dest='wizard', help='Run the configuration wizard.')
        parser.add_argument('--configure-modules', action='store_true',
                          dest='mod_wizard', help=(
                              'Run the configuration wizard, but only for the '
                              'module configuration options.'))
        parser.add_argument('--configure-database', action='store_true',
                          dest='db_wizard', help=(
                              'Run the configuration wizard, but only for the '
                              'database configuration options.'))
        opts = parser.parse_args()

        try:
            if os.getuid() == 0 or os.geteuid() == 0:
                stderr('Error: Do not run UnivBot with root privileges.')
                sys.exit(1)
        except AttributeError:
            # Windows don't have os.getuid/os.geteuid
            pass

        if opts.wizard:
            wizard('all', opts.config)
            return
        elif opts.mod_wizard:
            wizard('mod', opts.config)
            return
        elif opts.db_wizard:
            wizard('db', opts.config)
            return

        if opts.list_configs:
            configs = enumerate_configs()
            print('Config files in ~/.UnvBot:')
            if len(configs) is 0:
                print('\tNone found')
            else:
                for config in configs:
                    print('\t%s' % config)
            print('-------------------------')
            return

        config_name = opts.config or 'default'

        configpath = find_config(config_name)
        if not os.path.isfile(configpath):
            print("Welcome to UnivBot!\nI can't seem to find the configuration file, so let's generate it!\n")
            if not configpath.endswith('.cfg'):
                configpath = configpath + '.cfg'
            create_config(configpath)
            configpath = find_config(config_name)
        try:
            config_module = Config(configpath)
        except ConfigurationError as e:
            stderr(e)
            sys.exit(2)

        if config_module.core.not_configured:
            stderr('Bot is not configured, can\'t start')
            # exit with code 2 to prevent auto restart on fail by systemd
            sys.exit(2)

        if not config_module.has_option('core', 'homedir'):
            config_module.dotdir = homedir
            config_module.homedir = homedir
        else:
            homedir = config_module.core.homedir
            config_module.dotdir = config_module.core.homedir

        if not config_module.core.logdir:
            config_module.core.logdir = os.path.join(homedir, 'logs')
        logfile = os.path.os.path.join(config_module.logdir, 'stdio.log')
        if not os.path.isdir(config_module.logdir):
            os.mkdir(config_module.logdir)

        config_module.exit_on_error = opts.exit_on_error
        config_module._is_deamonized = opts.deamonize

        sys.stderr = tools.OutputRedirect(logfile, True, opts.quiet)
        sys.stdout = tools.OutputRedirect(logfile, False, opts.quiet)

        #Handle --quit, --kill and saving the PID to file
        pid_dir = config_module.core.pid_dir or homedir
        if opts.config is None:
            pid_file_path = os.path.join(pid_dir, 'UnivBot.pid')
        else:
            basename = os.path.basename(opts.config)
            if basename.endswith('.cfg'):
                basename = basename[:-4]
            pid_file_path = os.path.join(pid_dir, 'UnivBot-%s.pid' % basename)
        if os.path.isfile(pid_file_path):
            pid_file = open(pid_file_path, 'r')
            old_pid = int(pid_file.read())
            pid_file.close()
            if tools.check_pid(old_pid):
                if not opts.quit and not opts.kill:
                    stderr('There\'s already a UnivBot instance running with this config file')
                    stderr('Try using the --quit or the --kill options')
                    sys.exit(1)
                elif opts.kill:
                    stderr('Killing the UnivBot')
                    os.kill(old_pid, signal.SIGKILL)
                    sys.exit(0)
                elif opts.quit:
                    stderr('Signaling UnivBot to stop gracefully')
                    if hasattr(signal, 'SIGUSR1'):
                        os.kill(old_pid, signal.SIGUSR1)
                    else:
                        os.kill(old_pid, signal.SIGTERM)
                    sys.exit(0)
            elif not tools.check_pid(old_pid) and (opts.kill or opts.quit):
                stderr('UnivBot is not running!')
                sys.exit(1)
        elif opts.quit or opts.kill:
            stderr('UnivBot is not running!')
            sys.exit(1)
        if opts.deamonize:
            child_pid = os.fork()
            if child_pid is not 0:
                sys.exit()
        pid_file = open(pid_file_path, 'w')
        pid_file.write(str(os.getpid()))
        pid_file.close()
        config_module.pid_file_path = pid_file_path

        # Step Five: Initialise And Run UnivBot
        run(config_module)
    except KeyboardInterrupt:
        print("\n\nInterrupted")
        os._exit(1)
if __name__ == '__main__':
    main()
